messenger.rst

```diff
@@ -654,168 +1171,161 @@ This middleware can be configured to define the entity manager to use:
-            ->setAbstract(true)
-            // Main dependencies are defined by the parent definitions.
-            // Arguments provided in the middleware config will be appended on the child definition.
-            ->setArguments([new Reference('doctrine')]);
+    use Symfony\Component\Messenger\Envelope;
+    use Symfony\Component\Messenger\MessageBusInterface;
+    use Symfony\Component\Messenger\Stamp\DelayStamp;
 
-.. note::
+    public function index(MessageBusInterface $bus)
+    {
+        $bus->dispatch(new SmsNotification('...'), [
+            // wait 5 seconds before processing
+            new DelayStamp(5000)
+        ]);
 
-    Middleware factories only allow appending scalar and array arguments in config
-    (no references to other services). For most advanced use-cases, register a
-    concrete definition of the middleware manually and use its id.
+        // or explicitly create an Envelope
+        $bus->dispatch(new Envelope(new SmsNotification('...'), [
+            new DelayStamp(5000)
+        ]));
 
-Your own Transport
-------------------
+        // ...
+    }
 
-Once you have written your transport's sender and receiver, you can register your
-transport factory to be able to use it via a DSN in the Symfony application.
+Internally, each message is wrapped in an ``Envelope``, which holds the message
+and stamps. You can create this manually or allow the message bus to do it. There
+are a variety of different stamps for different purposes and they're used internally
+to track information about a message - like the message bus that's handling it
+or if it's being retried after failure.
 
-Create your Transport Factory
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Middleware
+~~~~~~~~~~
 
-You need to give FrameworkBundle the opportunity to create your transport from a
-DSN. You will need a transport factory::
+What happens when you dispatch a message to a message bus depends on its
+collection of middleware (and their order). By default, the middleware configured
+for each bus looks like this:
 
-    use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
-    use Symfony\Component\Messenger\Transport\Sender\SenderInterface;
-    use Symfony\Component\Messenger\Transport\TransportFactoryInterface;
-    use Symfony\Component\Messenger\Transport\TransportInterface;
+#. ``add_bus_name_stamp_middleware`` - adds a stamp to record which bus this
+   message was dispatched into.
 
-    class YourTransportFactory implements TransportFactoryInterface
-    {
-        public function createTransport(string $dsn, array $options): TransportInterface
-        {
-            return new YourTransport(/* ... */);
-        }
+#. ``dispatch_after_current_bus``- see :doc:`/messenger/message-recorder`.
 
-        public function supports(string $dsn, array $options): bool
-        {
-            return 0 === strpos($dsn, 'my-transport://');
-        }
-    }
+#. ``failed_message_processing_middleware`` - sends failed messages to the
```

Comment:
this information is wrong. the middleware does not send messages to failure transport (it's done via event listener), but it pretends to message was received from the original transport instead of the failure transport so that the from_transport config works